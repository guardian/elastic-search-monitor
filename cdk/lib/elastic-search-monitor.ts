import { GuScheduledLambda } from "@guardian/cdk";
import { GuAlarm } from "@guardian/cdk/lib/constructs/cloudwatch";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuVpc } from "@guardian/cdk/lib/constructs/ec2";
import type { App } from "aws-cdk-lib";
import { CfnParameter, Duration, Size } from "aws-cdk-lib";
import { ComparisonOperator, Metric } from "aws-cdk-lib/aws-cloudwatch";
import { SecurityGroup } from "aws-cdk-lib/aws-ec2";
import { Schedule } from "aws-cdk-lib/aws-events";
import { PolicyStatement } from "aws-cdk-lib/aws-iam";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { Topic } from "aws-cdk-lib/aws-sns";
import { EmailSubscription } from "aws-cdk-lib/aws-sns-subscriptions";

const app = "elastic-search-monitor";

export class ElasticSearchMonitor extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);
    const clusterSecurityGroupId = new CfnParameter(
     this,
     "ClusterSecurityGroup",
     {
       type: "AWS::EC2::SecurityGroup::Id",
       description:
        "A security group allowing the lambda to connect to Elasticsearch on port 9200 over TCP",
     }
    );
    const networkInterfacePolicy = new PolicyStatement({
      actions: [
        "ec2:CreateNetworkInterface",
        "ec2:DescribeNetworkInterfaces",
        "ec2:DeleteNetworkInterface",
      ],
      resources: ["*"],
    });
    const ec2DetectionPolicy = new PolicyStatement({
      actions: ["ec2:DescribeInstances"],
      resources: ["*"],
    });
    const pushMetricsPolicy = new PolicyStatement({
      actions: ["cloudwatch:PutMetricData"],
      resources: ["*"],
    });
    const additionalPolicies: PolicyStatement[] = [
      networkInterfacePolicy,
      ec2DetectionPolicy,
      pushMetricsPolicy,
    ];
    const scheduledLambda = new GuScheduledLambda(this, "ScheduledLambda", {
      app,
      fileName: "elastic-search-monitor.jar",
      environment: {
        TAG_QUERY_APP: "elk-es-master",
        TAG_QUERY_STACK: "deploy",
        CLUSTER_NAME: "elk",
      },
      description:
       "Monitors your Elasticsearch cluster and reports metrics to CloudWatch",
      handler: "com.gu.elasticsearchmonitor.Lambda::handler",
      monitoringConfiguration: {
        toleratedErrorPercentage: 99,
        numberOfEvaluationPeriodsAboveThresholdBeforeAlarm: 30,
        snsTopicName: "devx-alerts",
      },
      rules: [{ schedule: Schedule.rate(Duration.minutes(1)) }],
      runtime: Runtime.JAVA_11,
      // This lambda needs access to the Deploy Tools VPC so that it can talk to the ELK stack
      vpc: GuVpc.fromIdParameter(this, "vpc"),
      vpcSubnets: {
        subnets: GuVpc.subnetsFromParameter(this),
      },
      securityGroups: [
        SecurityGroup.fromSecurityGroupId(
         this,
         "ElasticsearchClusterSecurityGroup",
         clusterSecurityGroupId.valueAsString
        ),
      ],
    });
    additionalPolicies.map((policy) => scheduledLambda.addToRolePolicy(policy));

    const topicForElasticsearchAlerts = new Topic(this, "ElkAlertChannel", {
      displayName: `ELK alert channel for ${this.stage}`,
      topicName: `elk-alerts-${this.stage}`,
    });

    topicForElasticsearchAlerts.addSubscription(
     new EmailSubscription("devx.sec.ops@guardian.co.uk")
    );

    const clusterName = "elk";

    const metric = (metricName: string) => {
      return new Metric({
        metricName,
        namespace: `${this.stack}/${clusterName}`,
        dimensionsMap: {
          Cluster: clusterName,
        },
      });
    };

    const commonAlarmProps = {
      app,
      snsTopicName: topicForElasticsearchAlerts.topicName,
      period: Duration.minutes(1),
    };

    const lessThanAlarmProps = {
      ...commonAlarmProps,
      comparisonOperator: ComparisonOperator.LESS_THAN_THRESHOLD,
      statistic: "Minimum",
    };

    const greaterThanAlarmProps = {
      ...commonAlarmProps,
      comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
      statistic: "Maximum",
    };

    new GuAlarm(this, "DataNodeCountAlarm", {
      ...lessThanAlarmProps,
      alarmDescription: `Unexpected count of data nodes in ${this.stage}. Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY`,
      evaluationPeriods: 2,
      metric: metric("NumberOfDataNodes"),
      threshold: 5,
    });

    new GuAlarm(this, "MasterNodeCountAlarm", {
      ...lessThanAlarmProps,
      alarmDescription: `Unexpected count of master nodes in ${this.stage}. Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY`,
      evaluationPeriods: 2,
      metric: metric("NumberOfRespondingMastersNodes"),
      threshold: 3,
    });

    new GuAlarm(this, "ClusterStatusAlarm", {
      ...greaterThanAlarmProps,
      alarmDescription: `Unexpected cluster status in ${this.stage}. See cloudwatch metric value for current cluster status (Green = 0, Yellow = 1, Red = 2). Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY`,
      evaluationPeriods: 30, // Tolerate yellow status for 30 mins before sending an alert
      metric: metric("Status"),
      // Green = 0, Yellow = 1, Red = 2
      threshold: 0,
    });

    new GuAlarm(this, "RedClusterStatusAlarm", {
      ...greaterThanAlarmProps,
      alarmDescription: `Red cluster status in ${this.stage}. See cloudwatch metric value for current cluster status (Green = 0, Yellow = 1, Red = 2). Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY`,
      evaluationPeriods: 1,
      metric: metric("Status"),
      // Green = 0, Yellow = 1, Red = 2
      threshold: 1,
    });

    new GuAlarm(this, "DataNodeJvmHeapUsageAlarm", {
      ...greaterThanAlarmProps,
      alarmDescription: `A data node is using too much of its heap in ${this.stage}`,
      evaluationPeriods: 2,
      metric: metric("MaxJvmHeapUsage"),
      threshold: 85,
    });

    const lowStorageDescription = `A data node is running on less than 80% disk space.
    For more context and troubleshooting instructions, see the runbook:
    https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY/edit#heading=h.8h00c65wqmv0`;

    // See: https://aws.amazon.com/ec2/instance-types/i3en/ and https://www.google.com/search?q=5000gb+in+gib&oq=5000gb+in+gib
    const totalStorage = Size.gibibytes(4657);
    const twentyPercentDiskSpaceInBytes = Math.round(
     totalStorage.toBytes() * 0.2
    );

    new GuAlarm(this, "DataNodeLowStorageAlarm", {
      ...lessThanAlarmProps,
      alarmDescription: lowStorageDescription,
      evaluationPeriods: 2,
      metric: metric("MinAvailableDiskSpace"),
      threshold: twentyPercentDiskSpaceInBytes,
    });
  }
}
