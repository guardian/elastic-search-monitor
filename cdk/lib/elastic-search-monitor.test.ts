import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { ElasticSearchMonitor } from "./elastic-search-monitor";

describe("The ElasticSearchMonitor stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new ElasticSearchMonitor(app, "ElasticSearchMonitor", {
      stack: "deploy",
      stage: "TEST",
    });
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
