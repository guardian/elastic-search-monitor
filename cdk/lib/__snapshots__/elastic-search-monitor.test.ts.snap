// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ElasticSearchMonitor stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuVpcParameter",
      "GuSubnetListParameter",
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
      "GuAlarm",
      "GuAlarm",
      "GuAlarm",
      "GuAlarm",
      "GuAlarm",
      "GuAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "ClusterSecurityGroup": {
      "Description": "A security group allowing the lambda to connect to Elasticsearch on port 9200 over TCP",
      "Type": "AWS::EC2::SecurityGroup::Id",
    },
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "PrivateSubnets": {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "VpcId": {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within. Should NOT be the account default VPC.",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
  },
  "Resources": {
    "ClusterStatusAlarm75657749": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "Unexpected cluster status in TEST. See cloudwatch metric value for current cluster status (Green = 0, Yellow = 1, Red = 2). Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 30,
        "MetricName": "Status",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Maximum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "DataNodeCountAlarm0ECE9147": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "Unexpected count of data nodes in TEST. Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "NumberOfDataNodes",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Minimum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 4,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "DataNodeJvmHeapUsageAlarmDECB3926": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "A data node is using too much of its heap in TEST",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "MaxJvmHeapUsage",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Maximum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 85,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "DataNodeLowStorageAlarm72D42DBD": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "A data node is running on less than 80% disk space.
    For more context and troubleshooting instructions, see the runbook:
    https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY/edit#heading=h.8h00c65wqmv0",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "MinAvailableDiskSpace",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Minimum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 1000083134874,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ElkAlertChannel1915A337": {
      "Properties": {
        "DisplayName": "ELK alert channel for TEST",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "TopicName": "elk-alerts-TEST",
      },
      "Type": "AWS::SNS::Topic",
    },
    "ElkAlertChanneldevxsecopsguardiancouk276AF53D": {
      "Properties": {
        "Endpoint": "devx.sec.ops@guardian.co.uk",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "ElkAlertChannel1915A337",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "MasterNodeCountAlarmFE01B272": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "Unexpected count of master nodes in TEST. Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY",
        "ComparisonOperator": "LessThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 2,
        "MetricName": "NumberOfRespondingMastersNodes",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Minimum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 3,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "RedClusterStatusAlarm42644991": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Fn::GetAtt": [
                    "ElkAlertChannel1915A337",
                    "TopicName",
                  ],
                },
              ],
            ],
          },
        ],
        "AlarmDescription": "Red cluster status in TEST. See cloudwatch metric value for current cluster status (Green = 0, Yellow = 1, Red = 2). Runbook: https://docs.google.com/document/d/1PuEvL7L-CTV72Jx4OmiB3y5hlMmJR7Xz-YxYWAMiGdY",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": [
          {
            "Name": "Cluster",
            "Value": "elk",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Status",
        "Namespace": "deploy/elk",
        "Period": 60,
        "Statistic": "Maximum",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScheduledLambda68819433": {
      "DependsOn": [
        "ScheduledLambdaServiceRoleDefaultPolicy790AF47E",
        "ScheduledLambdaServiceRoleC946C6E8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/TEST/elastic-search-monitor/elastic-search-monitor.jar",
        },
        "Description": "Monitors your Elasticsearch cluster and reports metrics to CloudWatch",
        "Environment": {
          "Variables": {
            "APP": "elastic-search-monitor",
            "CLUSTER_NAME": "elk",
            "STACK": "deploy",
            "STAGE": "TEST",
            "TAG_QUERY_APP": "elk-es-master",
            "TAG_QUERY_STACK": "deploy",
          },
        },
        "Handler": "com.gu.elasticsearchmonitor.Lambda::handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "ScheduledLambdaServiceRoleC946C6E8",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "elastic-search-monitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 30,
        "VpcConfig": {
          "SecurityGroupIds": [
            {
              "Ref": "ClusterSecurityGroup",
            },
          ],
          "SubnetIds": {
            "Ref": "PrivateSubnets",
          },
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "ScheduledLambdaErrorPercentageAlarmForLambda6F1E8BCD": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":devx-alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "ScheduledLambda68819433",
              },
              " exceeded 99% error rate",
            ],
          ],
        },
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "High error percentage from ",
              {
                "Ref": "ScheduledLambda68819433",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 30,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "ScheduledLambda68819433",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "ScheduledLambda68819433",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "ScheduledLambda68819433",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Threshold": 99,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScheduledLambdaScheduledLambdarate1minute0064C7EC4": {
      "Properties": {
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "ScheduledLambda68819433",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "ScheduledLambdaScheduledLambdarate1minute0AllowEventRuleElasticSearchMonitorScheduledLambdaDFD7F6B32CFE3E7A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ScheduledLambda68819433",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "ScheduledLambdaScheduledLambdarate1minute0064C7EC4",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ScheduledLambdaServiceRoleC946C6E8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "elastic-search-monitor",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/elastic-search-monitor",
          },
          {
            "Key": "Stack",
            "Value": "deploy",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ScheduledLambdaServiceRoleDefaultPolicy790AF47E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/TEST/elastic-search-monitor/elastic-search-monitor.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/deploy/elastic-search-monitor",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/deploy/elastic-search-monitor/*",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ec2:CreateNetworkInterface",
                "ec2:DescribeNetworkInterfaces",
                "ec2:DeleteNetworkInterface",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "ec2:DescribeInstances",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ScheduledLambdaServiceRoleDefaultPolicy790AF47E",
        "Roles": [
          {
            "Ref": "ScheduledLambdaServiceRoleC946C6E8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
